python/tvm/meta_schedule/runner/local_runner.py

```diff
@@ -0,0 +1,367 @@
+            future = self.pool.submit(
+                LocalRunner._worker_func,
+                self.f_alloc_argument,
+                self.f_run_evaluator,
+                self.f_cleanup,
+                self.evaluator_config,
+                self.alloc_repeat,
+                str(runner_input.artifact_path),
+                str(runner_input.device_type),
+                tuple(arg_info.as_json() for arg_info in runner_input.args_info),
+            )
+            try:
+                result: List[float] = future.result()
+                error_message: str = None
+            except TimeoutError as exception:
+                result: List[float] = None
+                error_message: str = (
+                    f"LocalRunner: Timeout, killed after {self.timeout_sec} seconds\n"
+                )
+            except Exception as exception:  # pylint: disable=broad-except
+                result: List[float] = None
+                error_message: str = "LocalRunner: An exception occurred\n" + str(exception)
+            local_future = LocalRunnerFuture(result=result, error_message=error_message)
+            results.append(local_future)
+        return results
+
+    @staticmethod
+    def _worker_func(
+        _f_alloc_argument: Optional[str],
+        _f_run_evaluator: Optional[str],
+        _f_cleanup: Optional[str],
+        evaluator_config: EvaluatorConfig,
+        alloc_repeat: int,
+        artifact_path: str,
+        device_type: str,
+        args_info: T_ARG_INFO_JSON_OBJ_LIST,
+    ) -> List[float]:
+        f_alloc_argument: LocalRunner.T_ALLOC_ARGUMENT = get_global_func_with_default_on_worker(
+            _f_alloc_argument, default_alloc_argument
+        )
+        f_run_evaluator: LocalRunner.T_RUN_EVALUATOR = get_global_func_with_default_on_worker(
+            _f_run_evaluator, default_run_evaluator
+        )
+        f_cleanup: LocalRunner.T_CLEANUP = get_global_func_with_default_on_worker(
+            _f_cleanup, default_cleanup
+        )
+
+        @contextmanager
+        def resource_handler():
+            try:
+                yield
+            finally:
+                # Step 5. Clean up
+                f_cleanup()
+
+        with resource_handler():
+            # Step 1: create the local runtime module
+            rt_mod = tvm.runtime.load_module(artifact_path)
+            # Step 2: create the local device
+            device = tvm.runtime.device(dev_type=device_type, dev_id=0)
+            # Step 3: Allocate input arguments
+            repeated_args: List[T_ARGUMENT_LIST] = f_alloc_argument(
+                device,
+                args_info,
+                alloc_repeat,
+            )
+            # Step 4: Run time_evaluator
+            costs: List[float] = f_run_evaluator(
+                rt_mod,
+                device,
+                evaluator_config,
+                repeated_args,
+            )
+        return costs
+
+
+def default_alloc_argument(
```

Comment:
Is it the identical as `rpc_runner.default_alloc_argument`? If so, is it possible to separate in a different file and import into `local_runner` and `rpc_runner`?

Reply:
The only difference is that the `default_alloc_argument` for `local_runner` doesn't take `RPCSession` as an argument.

Reply:
I see. Thanks for the clarification! So probably we could merge the common part into a separate file (with `f_random_fill` as an extra argument), then call the method in both files. Same for `default_run_evaluator`.

Reply:
Agreed with this comment.

Reply:
This is done. Two duplicated functions are moved to `runner/utils.py`.